File: mobile/test/features/profile_tier1/presentation/utils/pricing_format_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:intl/date_symbol_data_local.dart';
import 'package:marketplace_mobile/features/profile_tier1/domain/entities/pricing.dart';
import 'package:marketplace_mobile/features/profile_tier1/domain/entities/pricing_kind.dart';
import 'package:marketplace_mobile/features/profile_tier1/presentation/utils/pricing_format.dart';

void main() {
  setUpAll(() async {
    await initializeDateFormatting();
  });

  group('formatPricing — daily rate', () {
    test('French locale uses /j suffix', () {
      const p = Pricing(
        kind: PricingKind.direct,
        type: PricingType.daily,
        minAmount: 50000,
        maxAmount: null,
        currency: 'EUR',
        note: '',
      );
      final out = formatPricing(p, locale: 'fr');
      expect(out, contains('500'));
      expect(out, endsWith('/j'));
    });

    test('English locale uses /day suffix', () {
      const p = Pricing(
        kind: PricingKind.direct,
        type: PricingType.daily,
        minAmount: 50000,
        maxAmount: null,
        currency: 'EUR',
        note: '',
      );
      final out = formatPricing(p, locale: 'en');
      expect(out, contains('500'));
      expect(out, endsWith('/day'));
    });
  });

  group('formatPricing — hourly rate', () {
    test('USD with French locale', () {
      const p = Pricing(
        kind: PricingKind.direct,
        type: PricingType.hourly,
        minAmount: 7500,
        maxAmount: null,
        currency: 'USD',
        note: '',
      );
      final out = formatPricing(p, locale: 'fr');
      expect(out, contains('75'));
      expect(out, endsWith('/h'));
    });
  });

  group('formatPricing — project range', () {
    test('formats both bounds with dash separator', () {
      const p = Pricing(
        kind: PricingKind.direct,
        type: PricingType.projectRange,
        minAmount: 1500000,
        maxAmount: 5000000,
        currency: 'EUR',
        note: '',
      );
      final out = formatPricing(p, locale: 'fr');
      expect(out, contains('–'));
      expect(out, contains('15'));
      expect(out, contains('50'));
    });

    test('falls back to "From X" when max is null', () {
      const p = Pricing(
        kind: PricingKind.direct,
        type: PricingType.projectRange,
        minAmount: 1500000,
        maxAmount: null,
        currency: 'EUR',
        note: '',
      );
      final en = formatPricing(p, locale: 'en');
      expect(en, contains('From'));

      final fr = formatPricing(p, locale: 'fr');
      expect(fr, contains('partir de'));
    });
  });

  group('formatPricing — project from', () {
    test('prefixes with "From" in English', () {
      const p = Pricing(
        kind: PricingKind.direct,
        type: PricingType.projectFrom,
        minAmount: 300000,
        maxAmount: null,
        currency: 'EUR',
        note: '',
      );
      final en = formatPricing(p, locale: 'en');
      expect(en, contains('From'));
      expect(en, contains('3'));
    });

    test('prefixes with "À partir de" in French', () {
      const p = Pricing(
        kind: PricingKind.direct,
        type: PricingType.projectFrom,
        minAmount: 300000,
        maxAmount: null,
        currency: 'EUR',
        note: '',
      );
      final fr = formatPricing(p, locale: 'fr');
      expect(fr, contains('partir de'));
    });
  });

  group('formatPricing — commission percent', () {
    test('basis points are converted to human percent', () {
      const p = Pricing(
        kind: PricingKind.referral,
        type: PricingType.commissionPct,
        minAmount: 550,
        maxAmount: null,
        currency: 'pct',
        note: '',
      );
      final fr = formatPricing(p, locale: 'fr');
      expect(fr, contains('5.5'));
      expect(fr, contains('%'));
    });

    test('range of percentages with integers', () {
      const p = Pricing(
        kind: PricingKind.referral,
        type: PricingType.commissionPct,
        minAmount: 500,
        maxAmount: 1500,
        currency: 'pct',
        note: '',
      );
      final out = formatPricing(p, locale: 'en');
      expect(out, contains('5'));
      expect(out, contains('15'));
      expect(out, contains('–'));
      expect(out, contains('%'));
    });
  });

  group('formatPricing — commission flat', () {
    test('French locale uses "/ deal" suffix', () {
      const p = Pricing(
        kind: PricingKind.referral,
        type: PricingType.commissionFlat,
        minAmount: 300000,
        maxAmount: null,
        currency: 'EUR',
        note: '',
      );
      final fr = formatPricing(p, locale: 'fr');
      expect(fr, contains('3'));
      expect(fr, contains('deal'));
    });

    test('English locale uses "per deal" suffix', () {
      const p = Pricing(
        kind: PricingKind.referral,
        type: PricingType.commissionFlat,
        minAmount: 300000,
        maxAmount: null,
        currency: 'EUR',
        note: '',
      );
      final en = formatPricing(p, locale: 'en');
      expect(en, contains('per deal'));
    });
  });

  group('formatPricingSummary', () {
    test('returns empty string for empty list', () {
      expect(formatPricingSummary(const [], locale: 'fr'), '');
    });

    test('joins direct + referral with a bullet separator', () {
      const direct = Pricing(
        kind: PricingKind.direct,
        type: PricingType.daily,
        minAmount: 50000,
        maxAmount: null,
        currency: 'EUR',
        note: '',
      );
      const referral = Pricing(
        kind: PricingKind.referral,
        type: PricingType.commissionPct,
        minAmount: 500,
        maxAmount: null,
        currency: 'pct',
        note: '',
      );
      final out = formatPricingSummary(
        [direct, referral],
        locale: 'en',
      );
      expect(out, contains('•'));
      expect(out, contains('/day'));
      expect(out, contains('%'));
    });
  });
}

File: web/src/features/provider/lib/__tests__/pricing-format.test.ts
import { describe, expect, it } from "vitest"
import { formatPricing } from "../pricing-format"
import type { Pricing } from "../../api/profile-api"

function row(overrides: Partial<Pricing>): Pricing {
  return {
    kind: "direct",
    type: "daily",
    min_amount: 50000,
    max_amount: null,
    currency: "EUR",
    note: "",
    negotiable: false,
    ...overrides,
  }
}

describe("formatPricing", () => {
  it("formats daily EUR in French", () => {
    const result = formatPricing(
      row({ type: "daily", min_amount: 50000, currency: "EUR" }),
      "fr",
    )
    expect(result).toContain("500")
    expect(result).toContain("€")
    expect(result).toContain("/j")
  })

  it("formats daily EUR in English", () => {
    const result = formatPricing(
      row({ type: "daily", min_amount: 50000, currency: "EUR" }),
      "en",
    )
    expect(result).toContain("500")
    expect(result).toContain("€")
    expect(result).toContain("/day")
  })

  it("formats hourly USD in English", () => {
    const result = formatPricing(
      row({ type: "hourly", min_amount: 7500, currency: "USD" }),
      "en",
    )
    expect(result).toContain("75")
    expect(result).toContain("$")
    expect(result).toContain("/hr")
  })

  it("formats project_from EUR with FR prefix", () => {
    const result = formatPricing(
      row({ type: "project_from", min_amount: 1000000, currency: "EUR" }),
      "fr",
    )
    expect(result).toMatch(/À partir de/)
    expect(result).toContain("10")
    expect(result).toContain("000")
    expect(result).toContain("€")
  })

  it("formats project_from EUR with EN prefix", () => {
    const result = formatPricing(
      row({ type: "project_from", min_amount: 1000000, currency: "EUR" }),
      "en",
    )
    expect(result).toMatch(/From/)
  })

  it("formats project_range EUR with both ends and single currency symbol on the max side", () => {
    const result = formatPricing(
      row({
        type: "project_range",
        min_amount: 1500000,
        max_amount: 5000000,
        currency: "EUR",
      }),
      "fr",
    )
    expect(result).toMatch(/15[\s\u202f\u00a0]?000\s?–\s?50[\s\u202f\u00a0]?000\s?€/)
  })

  it("formats commission_pct with trailing %", () => {
    const result = formatPricing(
      row({
        type: "commission_pct",
        min_amount: 500,
        max_amount: 1500,
        currency: "pct",
      }),
      "fr",
    )
    expect(result).toContain("5")
    expect(result).toContain("15")
    expect(result).toContain("%")
  })

  it("formats commission_pct without max as a single percent", () => {
    const result = formatPricing(
      row({
        type: "commission_pct",
        min_amount: 550,
        max_amount: null,
        currency: "pct",
      }),
      "fr",
    )
    expect(result).toContain("5,5")
    expect(result).toContain("%")
    expect(result).not.toContain("–")
  })

  // V1 pricing simplification: the referrer editor stores the single
  // user-entered percentage in BOTH min and max. formatPricing must
  // collapse the range when the two bounds are equal so the card
  // surfaces "10 % de commission" instead of "10 – 10 %".
  it("collapses commission_pct to a single headline when min equals max (V1)", () => {
    const resultFR = formatPricing(
      row({
        type: "commission_pct",
        min_amount: 1000,
        max_amount: 1000,
        currency: "pct",
      }),
      "fr",
    )
    expect(resultFR).toContain("10")
    expect(resultFR).toContain("%")
    expect(resultFR).toContain("commission")
    expect(resultFR).not.toContain("–")

    const resultEN = formatPricing(
      row({
        type: "commission_pct",
        min_amount: 1000,
        max_amount: 1000,
        currency: "pct",
      }),
      "en",
    )
    expect(resultEN).toContain("10")
    expect(resultEN).toContain("%")
    expect(resultEN).toContain("commission")
    expect(resultEN).not.toContain("–")
  })

  it("formats commission_flat EUR with 'per deal' suffix in English", () => {
    const result = formatPricing(
      row({
        type: "commission_flat",
        min_amount: 300000,
        currency: "EUR",
      }),
      "en",
    )
    expect(result).toContain("3,000")
    expect(result).toMatch(/per deal/)
  })

  it("formats commission_flat EUR with '/ deal' suffix in French", () => {
    const result = formatPricing(
      row({
        type: "commission_flat",
        min_amount: 300000,
        currency: "EUR",
      }),
      "fr",
    )
    expect(result).toMatch(/deal/)
  })

  it("handles GBP hourly rates in English", () => {
    const result = formatPricing(
      row({ type: "hourly", min_amount: 10000, currency: "GBP" }),
      "en",
    )
    expect(result).toContain("£")
    expect(result).toContain("100")
  })

  it("handles CAD daily rates in French", () => {
    const result = formatPricing(
      row({ type: "daily", min_amount: 80000, currency: "CAD" }),
      "fr",
    )
    expect(result).toContain("800")
    expect(result).toContain("/j")
  })
})
